orttraining/orttraining/core/optimizer/memory_optimizer/memory_insight.h

```diff
@@ -0,0 +1,108 @@
+                                          node_index_to_its_order_in_topological_sort_map,
+                                      ptrdiff_t& yield_op_order_in_topological_sort,
+                                      InlinedHashMap<const Node*, InlinedVector<size_t>>& candidate_output_args_map,
+                                      MemoryOptimizationPlanner& mem_opt_stats);
+
+/**
+ * @brief From the optimization plans, generate the memory optimization statistics table containing many MemoryRecords,
+ * each represents one node cluster id.
+ *
+ * @param memory_opt_planner The optimization planner to get optimization plans.
+ * @param generated_records Returns the generated memory optimization statistics table.
+ * @param node_to_apply_contexts_map Optional. If provided, we will append the optimization applying information
+ * (for example, how many are actually applied) to each MemoryRecord.
+ */
+void GetMemoryRecordsGroupedByNodeClusterId(const MemoryOptimizationPlanner& memory_opt_planner,
+                                            std::vector<std::pair<std::string, MemoryRecord>>& generated_records,
+                                            const NodeToClusterApplyContextMap&
+                                                node_to_apply_contexts_map);
+
+/**
+ * @brief Serialize the memory optimization statistics table to a string.
+ *
+ * @param records_grouped_by_node_cluster_id The memory optimization statistics table.
+ * @param user_config Optional. If provided, we will append the user configuration to the serialized string.
+ * @return std::string
+ */
+std::string SerializeMemoryRecords(const std::vector<std::pair<std::string, MemoryRecord>>&
+                                       records_grouped_by_node_cluster_id,
+                                   const std::string user_config = "");
+
+/**
+ * @brief A public API exposed to retrieve the memory optimization statistics table, given a graph.
+ *
+ * If possible, session's allocation plans and execution plan will also be available to help the analysis.
+ *
+ * @param graph The graph to analyze.
+ * @param memory_optimization_config The user configuration to control the memory optimization.
+ * @param recompute_probe_level The level to control allowed operations during recomputable subgraph detecting.
+ * @param logger Logger.
+ * @param ortvalue_name_to_idx_map Optional. If provided, we will use it to map ort value name to index.
+ * @param p_seq_exec_plan Optional. If provided, we will use it to get allocation plans.
+ * @return std::string
+ */
+std::string GetSerializedORTModuleMemoryStat(const Graph& graph,
+                                             const std::string& memory_optimization_config,
+                                             const std::string recompute_probe_level,
+                                             const logging::Logger& logger,
+                                             const OrtValueNameIdxMap& ortvalue_name_to_idx_map = {},
```

Comment:
>const OrtValueNameIdxMap& ortvalue_name_to_idx_map = {} [](http://example.com/codeflow?start=45&length=55)

We use nullptr for optional arguments. Constructing maps for doing nothing is expensive.

Reply:
Thanks! Fixed it by using a pointer.